activation/validation.go

```diff
@@ -62,31 +60,54 @@ func (v *Validator) NIPost(nodeId types.NodeID, commitmentAtxId types.ATXID, nip
 		return 0, err
 	}
 
+	if err := validateMerkleProof(expectedChallenge[:], &nipost.Membership); err != nil {
+		return 0, fmt.Errorf("invalid membership proof %w", err)
+	}
+
+	if err := v.Post(nodeId, commitmentAtxId, nipost.Post, nipost.PostMetadata, numUnits, opts...); err != nil {
+		return 0, fmt.Errorf("invalid Post: %v", err)
+	}
+
 	var ref types.PoetProofRef
 	copy(ref[:], nipost.PostMetadata.Challenge)
-	proof, err := v.poetDb.GetProof(ref)
+	proof, statement, err := v.poetDb.GetProof(ref)
 	if err != nil {
 		return 0, fmt.Errorf("poet proof is not available %x: %w", nipost.PostMetadata.Challenge, err)
 	}
-
-	if !contains(proof, nipost.Challenge.Bytes()) {
-		return 0, fmt.Errorf("challenge is not included in the proof %x", nipost.PostMetadata.Challenge)
-	}
-
-	if err := v.Post(nodeId, commitmentAtxId, nipost.Post, nipost.PostMetadata, numUnits, opts...); err != nil {
-		return 0, fmt.Errorf("invalid Post: %v", err)
+	if !bytes.Equal(statement[:], nipost.Membership.Root[:]) {
+		return 0, fmt.Errorf("invalid membership proof - root doesn't match poet proof statement; expected: %x, given: %x", statement[:], nipost.Membership.Root[:])
 	}
 
 	return proof.LeafCount, nil
 }
 
-func contains(proof *types.PoetProof, member []byte) bool {
-	for _, part := range proof.Members {
-		if bytes.Equal(part[:], member) {
-			return true
+func validateMerkleProof(leaf []byte, proof *types.MerkleProof) error {
+	nodes := make([][]byte, 0, len(proof.Nodes))
+	for _, n := range proof.Nodes {
+		nodes = append(nodes, n.Bytes())
+	}
+	ok, err := merkle.ValidatePartialTree(
+		[]uint64{proof.LeafIndex},
+		[][]byte{leaf},
+		nodes, proof.Root.Bytes(),
```

Comment:
i think it is better to pass nipost.Statement directly here and drop proof.Root field, and that validation